README.md
# Earnings Ticker

Elevate your motivation and seize the day with **Earnings Ticker**, an extension that calculates your real-time earnings based on your annual USD salary and work hours. Simply input your salary and work schedule, and marvel as your income accumulates before your eyes during your workday.

## Features

**Real-Time Earnings Display:** Keep track of your hard-earned money as you work. Your earnings update in real time, providing a constant stream of motivation.

**Customizable Work Hours:** Specify your daily work start and end times to get the most accurate earnings data tailored to your schedule.

**Unobtrusive Yet Accessible:** The extension sits quietly in your status bar, ensuring it doesn't distract you from your main tasks.

## How-To

1. Open the **Command Palette** on Visual Studio Code.
2. Enter `Start Earnings Ticker` and follow the prompts.
3. Enter `Cancel Earnings Ticker` to cancel.

## Default Settings

- The extension assumes 261 calendar workdays (removing all weekends).
- The extension assumes 26 days off to account for holidays, sick days, and vacation days.

## Customizable Settings

- `earningsTicker.numberOfAnnualWorkdays`: Configure the number of workdays in your calendar year.

- `earningsTicker.numberOfDaysOffWork`: Configure the number of days taken off from work.

## Known Issues

- Does not account for odd working hours.
- Earnings accumulation does not persist across workspaces.

## To-Do

- [ ] Make earnings accumulation persist across workspaces.


CHANGELOG.md
# Change Log

All notable changes to the "earnings-ticker" extension will be documented in this file.

## [Unreleased]

- Initial release

## 0.0.2 - 28 Oct 2023

- Re-release

## 0.0.3 - 28 Oct 2023

### Added:

- New logo

## 0.0.4 - 30 Oct 2023

### Added:

- Configuration settings

## 0.0.5 - 30 Oct 2023

### Added:

- Additional information to README file

## 0.0.5 - 31 Oct 2023

- Small refactor

## 0.0.8 - 27 Nov 2023

### Added:

- Support for foreign currencies outside of the US! 🎉


src/test/suite/index.ts
import * as path from 'path';
import Mocha from 'mocha';
import { glob } from 'glob';

export async function run(): Promise<void> {
  // Create the mocha test
  const mocha = new Mocha({
    ui: 'tdd',
    color: true
  });

  const testsRoot = path.resolve(__dirname, '..');
  const files = await glob('**/**.test.js', { cwd: testsRoot });

  // Add files to the test suite
  files.forEach(f => mocha.addFile(path.resolve(testsRoot, f)));

  try {
    return new Promise<void>((c, e) => {
      // Run the mocha test
      mocha.run(failures => {
        if (failures > 0) {
          e(new Error(`${failures} tests failed.`));
        } else {
          c();
        }
      });
    });
  } catch (err) {
    console.error(err);
  }
}


src/test/suite/extension.test.ts
import * as assert from 'assert';
import * as vscode from 'vscode';
import * as myExtension from '../../extension';

suite('Extension Test Suite', () => {
  before(() => {
    vscode.window.showInformationMessage('Starting all tests!');
  });

  after(() => {
    vscode.window.showInformationMessage('All tests done!');
  });

  test('Sample test', () => {});
});


src/extension.ts
import * as vscode from 'vscode';

enum Commands {
  startEarningsTicker = 'extension.startEarningsTicker',
  cancelEarningsTicker = 'extension.cancelEarningsTicker'
}
type UserInputs = string[];

export default class EarningsTicker {
  private currency: string = 'USD';
  private annualSalary: number = 0;
  private startTime: Date | null = null;
  private endTime: Date | null = null;
  private intervalId: NodeJS.Timeout | null = null;
  private statusBar: vscode.StatusBarItem | null = null;
  private workdayActive = false;
  private deactivationTimeoutSet = false;
  private numberOfAnnualWorkdays: number = vscode.workspace
    .getConfiguration('earningsTicker')
    .get<number>('numberOfAnnualWorkdays', 261);
  private numberOfDaysOffWork: number = vscode.workspace
    .getConfiguration('earningsTicker')
    .get<number>('numberOfDaysOffWork', 26);
  private daysActivelyWorking: number =
    this.numberOfAnnualWorkdays - this.numberOfDaysOffWork;

  constructor(private context: vscode.ExtensionContext) {}

  public activate() {
    vscode.workspace.onDidChangeConfiguration(e => {
      if (
        e.affectsConfiguration('earningsTicker.numberOfAnnualWorkdays') ||
        e.affectsConfiguration('earningsTicker.numberOfDaysOffWork')
      ) {
        this.reInitialize();
      }
    });

    this.loadState();

    this.context.subscriptions.push(
      vscode.commands.registerCommand(Commands.startEarningsTicker, () =>
        this.startEarningsTicker()
      )
    );

    this.context.subscriptions.push(
      vscode.commands.registerCommand(Commands.cancelEarningsTicker, () =>
        this.deactivate()
      )
    );
  }

  public deactivate() {
    if (this.intervalId) {
      clearInterval(this.intervalId);
    }

    if (this.statusBar) {
      this.statusBar.hide();
      this.statusBar.dispose();
    }

    this.resetState();

    vscode.window.showInformationMessage('Earnings Ticker has been canceled.');
  }

  private reInitialize() {
    this.updateConfigSettings();
    this.deactivate();
    this.loadState();
    this.startEarningsTicker();
  }

  private async startEarningsTicker() {
    const userInputs = await this.getUserInputs();

    if (!userInputs) {
      this.deactivate();
    } else if (!(await this.validateAndSetInputs(userInputs))) {
      this.deactivate();
    } else {
      this.updateStatusBar(this.annualSalary / this.daysActivelyWorking);
    }
  }

  private loadState() {
    const storedSalary = this.context.globalState.get<number>(
      'annualSalary',
      0
    );

    const storedStartTime = this.context.globalState.get<string>(
      'startTime',
      new Date().toISOString()
    );

    const storedEndTime = this.context.globalState.get<string>(
      'endTime',
      new Date().toISOString()
    );

    this.annualSalary = storedSalary;
    this.startTime = new Date(storedStartTime);
    this.endTime = new Date(storedEndTime);
  }

  private async updateState(
    annualSalary: number,
    startTime: Date,
    endTime: Date
  ): Promise<void> {
    try {
      await Promise.all([
        this.context.globalState.update('annualSalary', annualSalary),
        this.context.globalState.update('startTime', startTime?.toISOString()),
        this.context.globalState.update('endTime', endTime?.toISOString())
      ]);
    } catch (error: unknown) {
      if (error instanceof Error) {
        this.showErrorMessage(`Failed to update state: ${error.message}`);
      }
    }
  }

  private resetState() {
    this.annualSalary = 0;
    this.startTime = null;
    this.endTime = null;
    this.intervalId = null;
    this.statusBar = null;
    this.workdayActive = false;
    this.deactivationTimeoutSet = false;
  }

  private updateConfigSettings() {
    this.numberOfAnnualWorkdays = vscode.workspace
      .getConfiguration('earningsTicker')
      .get<number>('numberOfAnnualWorkdays', 261);
    this.numberOfDaysOffWork = vscode.workspace
      .getConfiguration('earningsTicker')
      .get<number>('numberOfDaysOffWork', 26);
    this.daysActivelyWorking =
      this.numberOfAnnualWorkdays - this.numberOfDaysOffWork;
  }

  private formatCurrency(value: number, currency = 'USD') {
    return new Intl.NumberFormat('en-US', {
      style: 'currency',
      currency: currency
    }).format(value);
  }

  private async getUserInputs(): Promise<UserInputs | null> {
    const currencyInput = await this.createPrompt(
      'Enter your desired currency (e.g., USD, EUR, GBP)'
    );

    if (currencyInput === undefined || currencyInput === '') {
      return null;
    }

    const salaryInput = await this.createPrompt(
      'Enter your annual salary (NUMBERS ONLY):'
    );

    if (salaryInput === undefined || salaryInput === '') {
      return null;
    }

    const startTimeInput = await this.createPrompt(
      'Enter your shift start time (HH:MM AM/PM):'
    );

    if (startTimeInput === undefined || startTimeInput === '') {
      return null;
    }

    const endTimeInput = await this.createPrompt(
      'Enter your shift end time (HH:MM AM/PM):'
    );

    if (endTimeInput === undefined || endTimeInput === '') {
      return null;
    }

    if (salaryInput && startTimeInput && endTimeInput) {
      return [currencyInput, salaryInput, startTimeInput, endTimeInput];
    } else {
      return null;
    }
  }

  private async createPrompt(prompt: string) {
    return await vscode.window.showInputBox({ prompt });
  }

  private parseSalary(salaryInput: string): number | null {
    const parsedSalary = parseFloat(salaryInput);
    if (parsedSalary <= 0 || !Boolean(parsedSalary)) {
      this.showErrorMessage(
        'Invalid salary amount. Please enter a positive number.'
      );
      return null;
    }
    return parsedSalary;
  }

  private validateTimeFormat(time: string): boolean {
    const timeRegex = /^([0-9]|0[1-9]|1[0-2]):[0-5][0-9]\s?(?:AM|PM|am|pm)$/;
    return timeRegex.test(time);
  }

  private convertTo24Hour(time: string): [number, number] {
    const [hour, minute] = time.split(/[:\s]/, 2);
    let meridian = time.match(/AM|PM|am|pm/gi);
    let hour24 = parseInt(hour);

    // Convert PM hours to 24-hour format, except for 12 PM
    if (meridian && meridian[0].toUpperCase() === 'PM' && hour !== '12') {
      hour24 += 12;
    }

    // Convert 12 AM to 0
    if (hour === '12' && meridian && meridian[0].toUpperCase() === 'AM') {
      hour24 = 0;
    }

    return [hour24, parseInt(minute)];
  }

  private async validateAndSetInputs(userInputs: UserInputs): Promise<boolean> {
    const [currencyInput, salaryInput, startTimeInput, endTimeInput] =
      userInputs;

    const parsedSalary = this.parseSalary(salaryInput);

    if (!parsedSalary) {
      return false;
    }

    if (
      !this.validateTimeFormat(startTimeInput) ||
      !this.validateTimeFormat(endTimeInput)
    ) {
      this.showErrorMessage('Invalid time format. Use HH:MM AM/PM.');
      return false;
    }

    this.annualSalary = parsedSalary;
    this.currency = currencyInput;

    const currentDate = new Date();
    this.startTime = new Date(currentDate);
    this.endTime = new Date(currentDate);

    const [startHour, startMinute] = this.convertTo24Hour(startTimeInput);
    const [endHour, endMinute] = this.convertTo24Hour(endTimeInput);

    this.startTime.setHours(startHour, startMinute, 0, 0);
    this.endTime.setHours(endHour, endMinute, 0, 0);

    // Ensure the start time is before the end time
    if (this.endTime <= this.startTime) {
      this.showErrorMessage('Start time must be before the end time.');
      return false;
    }

    try {
      await this.updateState(this.annualSalary, this.startTime, this.endTime);
      return true;
    } catch (error: unknown) {
      if (error instanceof Error) {
        this.showErrorMessage(`An error occurred. ${error.message}`);
      } else {
        this.showErrorMessage('An unknown error occurred.');
      }
      return false;
    }
  }

  private updateStatusBar(dailySalary: number) {
    if (!this.statusBar) {
      this.statusBar = vscode.window.createStatusBarItem(
        vscode.StatusBarAlignment.Right,
        100
      );
      this.statusBar.text = '';
      this.statusBar.color = '#39FF14';
      this.statusBar.show();
    }

    if (this.intervalId) {
      clearInterval(this.intervalId);
    }

    this.intervalId = setInterval(() => {
      if (this.startTime && this.endTime && dailySalary && this.statusBar) {
        const now = new Date();

        if (now >= this.startTime && now <= this.endTime) {
          this.workdayActive = true;
          const elapsedMs = now.getTime() - this.startTime.getTime();
          const totalWorkMs = this.endTime.getTime() - this.startTime.getTime();
          const earned = (elapsedMs / totalWorkMs) * dailySalary;

          this.statusBar.text = this.formatCurrency(earned, this.currency);
        } else if (this.workdayActive && !this.deactivationTimeoutSet) {
          this.deactivationTimeoutSet = true;
          this.workdayActive = false;

          this.statusBar.text = `Congrats! You Earned ${this.formatCurrency(
            dailySalary,
            this.currency
          )}`;

          setTimeout(() => {
            this.deactivate();
          }, 10 * 60 * 1000); // 10 minutes in milliseconds

          if (this.intervalId) {
            clearInterval(this.intervalId);
          }
        } else {
          this.statusBar.text = `Congrats! You Earned ${this.formatCurrency(
            dailySalary,
            this.currency
          )}`;
          setTimeout(() => {
            this.deactivate();
          }, 2 * 60 * 5000); // 2 minutes in milliseconds

          if (this.intervalId) {
            clearInterval(this.intervalId);
          }
        }
      }
    }, 1000);
  }

  private showErrorMessage(message: string) {
    vscode.window.showErrorMessage(message);
  }
}

let ticker: EarningsTicker;

export function activate(context: vscode.ExtensionContext) {
  ticker = new EarningsTicker(context);
  ticker.activate();
}

export function deactivate() {
  ticker.deactivate();
}
